File: apps/mobile/lib/features/assistant/assistant_blocks.dart
/// 小爱 blocks / 时间轴 / 静态结构（对齐 Web `assistant_blocks.ts`）。
library;

import 'assistant_format.dart';

class TimelineNode {
  const TimelineNode({
    required this.year,
    required this.label,
    this.note,
  });

  final String year;
  final String label;
  final String? note;

  factory TimelineNode.fromJson(Map<String, dynamic> j) => TimelineNode(
        year: (j['year'] ?? '') as String,
        label: (j['label'] ?? j['year'] ?? '') as String,
        note: j['note'] as String?,
      );
}

class StructureAsset {
  const StructureAsset({
    required this.kind,
    required this.id,
    required this.label,
    this.subtitle,
    this.href,
    this.nodes = const [],
  });

  final String kind;
  final String id;
  final String label;
  final String? subtitle;
  final String? href;
  final List<TimelineNode> nodes;

  factory StructureAsset.fromJson(Map<String, dynamic> j) => StructureAsset(
        kind: (j['kind'] ?? '') as String,
        id: (j['id'] ?? '') as String,
        label: (j['label'] ?? '') as String,
        subtitle: j['subtitle'] as String?,
        href: j['href'] as String?,
        nodes: ((j['nodes'] ?? []) as List)
            .map((e) => TimelineNode.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

const _timelineSectionTitles = {
  '时间线',
  '年代脉络',
  '历史脉络',
  '人物生平',
  '年代',
};
final _timelineItemRe = RegExp(
  r'^\s*(?:[-*•]|\d+[.)、])\s+\*\*(.+?)\*\*\s*(.*)$',
);

List<TimelineNode> parseTimelineNodes(String text) {
  final body = bodyText(text).trim();
  if (body.isEmpty) return const [];
  final titles = _timelineSectionTitles.join('|');
  final sectionRe = RegExp('(?:^|\\n)###\\s+(?:$titles)\\s*\\n', multiLine: true);
  final m = sectionRe.firstMatch(body);
  if (m == null) return const [];
  final tail = body.substring(m.end);
  final nextIdx = tail.indexOf(RegExp(r'\n###\s+'));
  final chunk = nextIdx >= 0 ? tail.substring(0, nextIdx) : tail;
  final nodes = <TimelineNode>[];
  for (final line in chunk.split('\n')) {
    final mm = _timelineItemRe.firstMatch(line.trim());
    if (mm == null) continue;
    final year = mm.group(1)!.trim();
    final note = (mm.group(2) ?? '').trim();
    nodes.add(TimelineNode(year: year, label: note.isEmpty ? year : note, note: note));
  }
  return nodes.length > 8 ? nodes.sublist(0, 8) : nodes;
}

Set<String> streamingWrittenSections(
  String text,
  List<({String id, String title})> sections,
) {
  final written = <String>{};
  final body = bodyText(text);
  for (final sec in sections) {
    final escaped = RegExp.escape(sec.title);
    if (RegExp('(?:^|\\n)###\\s+$escaped\\s*\\n+\\S', multiLine: true)
        .hasMatch(body)) {
      written.add(sec.id);
    }
  }
  return written;
}

String extractStudySheetCopyText(String text) {
  final body = bodyText(text);
  final m = RegExp(
    r'(?:^|\n)###\s*讨论问题\s*\n([\s\S]*?)(?=\n###\s+|$)',
  ).firstMatch(body);
  if (m == null) return '';
  final lines = m.group(1)!.split('\n');
  final out = <String>[];
  var i = 1;
  for (final ln in lines) {
    final q = ln.replaceFirst(RegExp(r'^\s*(?:[-*•]|\d+[.)、])\s+'), '').trim();
    if (q.isEmpty) continue;
    out.add('$i. $q');
    i += 1;
  }
  return out.join('\n\n');
}

File: apps/web/lib/assistant_blocks.ts
import { bodyText } from '@/lib/assistant_format';

export type TimelineNode = {
  year: string;
  label: string;
  note?: string;
};

export type AnswerBlock =
  | { type: 'lead'; text: string }
  | { type: 'timeline'; items: TimelineNode[] }
  | { type: 'list'; title: string; items: string[] }
  | { type: 'paragraph'; title: string; text: string };

export type StructureAsset = {
  kind: 'timeline' | 'diagram' | 'graph';
  id: string;
  label: string;
  subtitle?: string;
  href?: string;
  nodes?: TimelineNode[];
};

const TIMELINE_SECTION_TITLES = new Set([
  '时间线',
  '年代脉络',
  '历史脉络',
  '人物生平',
  '年代',
]);
const TIMELINE_ITEM_RE = /^\s*(?:[-*•]|\d+[.)、])\s+\*\*(.+?)\*\*\s*(.*)$/;

/** 从 ### 时间线 等小节反解析竖轴节点。 */
export function parseTimelineNodes(text: string): TimelineNode[] {
  const body = bodyText(text).trim();
  if (!body) return [];
  const titles = [...TIMELINE_SECTION_TITLES].join('|');
  const sectionRe = new RegExp(`(?:^|\\n)###\\s+(?:${titles})\\s*\\n`, 'm');
  const m = sectionRe.exec(body);
  if (!m) return [];
  const tail = body.slice(m.index + m[0].length);
  const nextIdx = tail.search(/\n###\s+/);
  const chunk = nextIdx >= 0 ? tail.slice(0, nextIdx) : tail;
  const nodes: TimelineNode[] = [];
  for (const line of chunk.split('\n')) {
    const mm = line.trim().match(TIMELINE_ITEM_RE);
    if (!mm?.[1]) continue;
    const year = mm[1].trim();
    const note = (mm[2] ?? '').trim();
    nodes.push({ year, label: note || year, note });
  }
  return nodes.slice(0, 8);
}

/** Markdown 正文反解析 blocks（与服务端 parse_answer_blocks 对齐）。 */
export function parseAnswerBlocks(text: string): {
  lead: string;
  blocks: AnswerBlock[];
  timeline: TimelineNode[];
} {
  const body = bodyText(text);
  const leadM = body.match(/(?:^|\n)###\s*(?:摘要|本章概览|卷概览)\s*\n+([^\n#]+)/);
  const lead = leadM?.[1]?.trim() ?? '';
  const blocks: AnswerBlock[] = [];
  if (lead) blocks.push({ type: 'lead', text: lead });
  const timeline = parseTimelineNodes(body);
  if (timeline.length) blocks.push({ type: 'timeline', items: timeline });
  return { lead, blocks, timeline };
}

/** 流式阶段：哪些小节标题后已有正文。 */
export function streamingWrittenSections(
  text: string,
  sections: { id: string; title: string }[],
): Set<string> {
  const written = new Set<string>();
  const body = bodyText(text);
  for (const sec of sections) {
    const re = new RegExp(
      `(?:^|\\n)###\\s+${sec.title.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\s*\\n+\\S`,
      'm',
    );
    if (re.test(body)) written.add(sec.id);
  }
  return written;
}

/** 查经/讲道 sheet：提取可复制的讨论题块。 */
export function extractStudySheetCopyText(text: string): string {
  const body = bodyText(text);
  const m = body.match(/(?:^|\n)###\s*讨论问题\s*\n([\s\S]*?)(?=\n###\s+|$)/);
  if (!m?.[1]) return '';
  return m[1]
    .split('\n')
    .map((ln) => ln.replace(/^\s*(?:[-*•]|\d+[.)、])\s+/, '').trim())
    .filter(Boolean)
    .map((q, i) => `${i + 1}. ${q}`)
    .join('\n\n');
}
